src/bsearch-dictionary.hpp

```diff
@@ -64,31 +64,35 @@ struct heap
 
 class BinarySearchDictionary : public Dictionary
 {
-  // Add any member variables you need
-  bool sorted;
-  heap * h;
-
 public:
   // Constructor
   BinarySearchDictionary();
   ~BinarySearchDictionary();
 
-  bool addRecord(QString key, int record);
+  bool add_record(const QString & key, int record) override;
 
   // Sort array using heap sort.
   void sort();
 
   int find(QString key);
 
   // Removes one element from the table
-  bool removeElement(QString key);
+  bool remove_element(const QString & key) override;
 
   // Returns all the elements in the table as an array of strings.
   // *n is the size of the table and it is returned by reference
-  QString * keys(int * n);
+  std::unique_ptr<QString[]> keys(size_t * n) override;
```

Comment:
I noticed that you've made use of std::unique alot more, what are the benefits of this?

Reply:
Good question! `std::unique_ptr` is really nice since it frees the resource automatically, and ensures that it cannot exist in multiple locations. The copy constructor/copy assignment operators are explicitly deleted in its implementation, allowing only for it to be created via r-value reference (`T &&`).

In easier English, it makes it so that you have to explicitly move the object to a different place or it gets cleaned up.

```c++
a = std::make_unique<int>(7);

{
  auto b = std::make_unique<int>(4);
  auto c = std::make_unique<int>(5);
  a = std::move(c);. // a is freed
}  // b is freed
// c exists, but is held by a
```

If that makes sense.